lib/gems/pending/util/miq_ftp_lib.rb

```diff
@@ -0,0 +1,84 @@
+require 'net/ftp'
+require 'logger'
+
+# Helper methods for net/ftp based classes and files.
+#
+# Will setup a `@ftp` attr_accessor to be used as the return value for
+# `.connect`, the main method being provided in this class.  Will also setup
+# logging if not already done for the particular class (that follows the
+# VmdbLogger conventions, and setup a `uri` attr_accessor if that doesn't
+# already exist.
+module MiqFtpLib
```

Comment:
I still think we should make this a class around FTP interaction with its own logger. Then we also don't have to be responsible for what methods are defined on the model (or other thing) using this class.

Reply:
@carbonin is this in response to this:

https://github.com/ManageIQ/manageiq-gems-pending/pull/360#issuecomment-417097055

or a comment on it's own?

Reply:
I think it's kind of both. I read that response a few days ago then re-read the whole mixin here and I'm still not comfortable with this being an included module.

I think there would be more value and stability in making this a standalone class than trying to fit this in as a mixin. The refactoring you would have to do in FileDepotFtp seems minimal and I think it would make this class *much* more stable and readable.

TLDR;
Make it a class, take a uri as a parameter to initialize, and give it a `#logger=` method with a null logger as a default. That should take care of all of the meta-programming in `.included` here.

Reply:
> I think there would be more value and stability...

Sure, maybe long term.  But as a reminder, this came about from your original comment here:

https://github.com/ManageIQ/manageiq/pull/17549#issuecomment-398062438

In which I was trying to meet halfway in regards to making this DRY.  But, to your comment that:

> The refactoring you would have to do in FileDepotFtp seems minimal...

I disagree entirely.  These methods are _**EXACTLY**_ as they are in `FileDepotFtp`, meaning there is next to zero risk on me doing what I originally had done in there as part of https://github.com/ManageIQ/manageiq/pull/17549 which was gut these methods from `FileDepotFTP` and replace them via `include MiqFtpLib` (previously [here](https://github.com/ManageIQ/manageiq/pull/17549/files#diff-aca27294178e2b711e7629bacef08fcbR4)).

With this suggestion, it is a re-write of `FileDepotFtp` effectively, because half of the methods I am not using here still have their own `_log` method that I would need to handle (or accept disparaging loggers for).

Also, `@ftp` is no longer part of the class, as it would be a reference to the new class, and so everything that isn't a part of `MiqFtpLib` as it is now that isn't removed from `FileDepotFtp` basically going to be re-written as a result.  The "specs" for `FileDepotFtp` (which is about 90% mocking, and 10% actually testing things), would also probably completely need to be re-written as a result.



* * * 

tl; dr  No, this would not be easier to write this as a class.  Unless you can convince "the powers that be" that this is worth said re-write that I didn't estimate for, I disagree with the suggestion.

Again [[ref]](https://github.com/ManageIQ/manageiq-gems-pending/pull/360#issuecomment-417097055), if the meta-programming is an issue, I can basically remove most of the `included` block (if not all of it), and just assume this is an "interface" module, and those methods inside of `included` can then required by the "including" class.  This goes for both the `uri` bit and the `_log`.

Reply:
Done.  I kept `attr_accessor :ftp` in since it has next to no meta-programming, but still tries to reach peak minimum LOC added, to appease @jrafanie ...